### PR TITLE
Split report.xslt to multiple files

### LINKED ISSUE #137: Split xslt template(s) into multiple files

### PR DESCRIPTION
**Pull request by Hebcak**
Closes #137

### CODE REVIEW

**Comment by paveljandejsek** [comment]
It seems like copying in XmlReporterFactory is not working correctly. Try to take a look locally at XmlReportIT, it might help to figure out whats wrong...

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 1]
```diff
@@ -157,12 +159,60 @@ private void copyResourceToReportDirectory(Resource resource) {
...
+
+    private void copyResourceFromJar(Resource resource, String targetFilePath, File targetFile, File jarFile) {
+    try{
+        try (JarFile jar = new JarFile(jarFile)) {
+            JarEntry entry = jar.getJarEntry(resource.getFilename());
+            if (entry.isDirectory()) {
+                jar.stream()
+                    .filter(file -> file.getName().startsWith(entry.getName()) && file.getName().endsWith(".xslt"))
+                    .forEach(file -> {
+                        String fileName = file.getName().replace(entry.getName(), "");
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Might be worth putting into separate method to improve readability

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 2]
```diff
@@ -157,12 +159,60 @@ private void copyResourceToReportDirectory(Resource resource) {
...
+                        File newFile = new File(targetFilePath + File.separator + fileName);
+                        try {
+                            FileUtils.copyInputStreamToFile(jar.getInputStream(file), newFile);
+                        } catch (IOException e) {
+                            log.error("Unable to get input stream for this file {}", targetFilePath + File.separator + fileName);
+                        }
+                    });
+            } else {
+                log.warn("File name: {}", entry.getName());
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Leftover from development? 🙂

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 3]
```diff
@@ -157,12 +159,60 @@ private void copyResourceToReportDirectory(Resource resource) {
     private void copyResourceToReportDirectory(Resource resource, String targetFilePath) {
         File targetFile = new File(targetFilePath);
         try {
-            FileUtils.copyInputStreamToFile(resource.getInputStream(), targetFile);
+            String resourcePath = String.valueOf(resource.getURI());
+            String jarPath = resourcePath.replace(("!/" + resource.getFilename()), "").replace("jar:file:", "");
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Just a thought - is this detection of Jar by URI enough? Cant we get `jarFile.isFile() == true` in some other cases? (just wondering)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 4]
```diff
@@ -157,12 +159,60 @@ private void copyResourceToReportDirectory(Resource resource) {
...
 
+    private void copyResourceFromFile(Resource resource, File targetFile) {
+        try {
+            if (resource.isFile() && resource.getFile().isDirectory()) {
+                FileUtils.copyDirectory(resource.getFile(), targetFile);
+            } else {
+                FileUtils.copyInputStreamToFile(resource.getInputStream(), targetFile);
+            }
+        }catch (IOException e){
+            log.error("{} is not a file!", resource.getFilename());
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Might be worth logging the exception as well (+ formatting 🙂)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 5]
```diff
@@ -157,12 +159,60 @@ private void copyResourceToReportDirectory(Resource resource) {
...
+                            log.error("Unable to get input stream for this file {}", targetFilePath + File.separator + fileName);
+                        }
+                    });
+            } else {
+                log.warn("File name: {}", entry.getName());
+
+                FileUtils.copyInputStreamToFile(jar.getInputStream(entry), targetFile);
+            }
+        }
+    } catch (IOException e) {
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Cant this catch be inside the first try? (So we might be able to get rid of those 2 try blocks and have just one?)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/jms/InterceptedMessageCreator.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package org.jbehavesupport.core.jms;
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/jms/InterceptedMessageCreator.java]: Might be worth putting it into `internal` package

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/jms/InterceptedMessageCreator.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import javax.jms.Message;
+import javax.jms.Session;
+
+public class InterceptedMessageCreator implements MessageCreator {
+
+    private MessageCreator messageCreator;
+
+    private JmsXmlReporterExtension jmsXmlReporterExtensions;
+
+    InterceptedMessageCreator(MessageCreator inputMessageCreator, JmsXmlReporterExtension reporter) {
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/jms/InterceptedMessageCreator.java]: Can be probably replaced by lombok constructor :slightly_smiling_face:

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/extension/JmsXmlReporterExtension.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                printBegin(writer, COLUMN_TAG);
+                printCData(writer, header);
+                printEnd(writer, COLUMN_TAG);
+
+                try {
+                    printBegin(writer, COLUMN_TAG);
+                    printCData(writer, message.getStringProperty(header));
+                    printEnd(writer, COLUMN_TAG);
+                } catch (JMSException e) {
+                    log.error("Unable to get the property: {}", header);
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/extension/JmsXmlReporterExtension.java]: Maybe we could print that inside the the tag? not sure, just something to think about if it makes sense or not in this case (+ maybe log the exception?)

File: jbehave-support-core/src/main/resources/functions.js [hunk 9]
```diff
@@ -18,6 +18,10 @@ function reportReady() {
         showHideAllInElement('#rest-calls-body');
     });
 
+    $('#expand-all-jms-messages').click(function () {
+            showHideAllInElement('#jms-messages-body');
+        });
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/resources/functions.js]: Nitpicking but seems like wrong indentation?

File: jbehave-support-core/src/main/resources/index.xslt [hunk 10]
```diff
@@ -229,8 +229,8 @@
                             Help
                         </a>
                     </div>
-                    <div class="col-4 text-muted text-center" id="userVersion" version="${project.version}">
-                        Created by jbehave-support reporter (build ${project.version})
+                    <div class="col-4 text-muted text-center" id="userVersion" version="1.0.8-SNAPSHOT">
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/resources/index.xslt]: Wrongly commited file :slightly_smiling_face:

File: jbehave-support-core/src/main/resources/report-generator/screenshotReport.xslt [hunk 11]
```diff
@@ -11,7 +9,8 @@
                 </xsl:attribute>
                 <div class="card-header">
                     <i class="fa fa-exchange" aria-hidden="true"></i>
-                    Step screen shots (<a href="#step-screen-shots" id="expand-all-step-screenshots">Toggle all screenshots</a>)
+                    Step screen shots (<a href="#step-screen-shots" id="expand-all-step-screenshots">Toggle all
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/resources/report-generator/screenshotReport.xslt]: Why the split to new line?

File: jbehave-support-core/src/main/resources/report-generator/serverLogXmlRep.xslt [hunk 12]
```diff
@@ -70,8 +67,11 @@
             </table>
 
             <div class="btn-group form-group align-self-center">
-                <a href="#sshlog-{$logNum}" data-toggle="collapse" class="btn btn-sm btn-outline-primary">Show/hide log contents</a>
-                <button type="button" class="btn btn-sm btn-outline-info btn-copy-clipboard" data-selector="#sshlog-{$logNum}" title="Copy to clipboard">
+                <a href="#sshlog-{$logNum}" data-toggle="collapse" class="btn btn-sm btn-outline-primary">Show/hide log
+                    contents
```
Comment (paveljandejsek) [inline jbehave-support-core/src/main/resources/report-generator/serverLogXmlRep.xslt]: Same with the newline :slightly_smiling_face:

File: jbehave-support-core/src/main/resources/report-generator/screenshotReport.xslt (hunk 11 above)
Comment (Hebcak) [reply jbehave-support-core/src/main/resources/report-generator/screenshotReport.xslt]: Probably just a misclick :-)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Comment by paveljandejsek** [comment]
Please rebase, ci and the reporting service were updated to support folders in reports....

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 13]
```diff
@@ -157,12 +159,60 @@ private void copyResourceToReportDirectory(Resource resource) {
...
+            if (entry.isDirectory()) {
+                jar.stream()
+                    .filter(file -> file.getName().startsWith(entry.getName()) && file.getName().endsWith(".xslt"))
+                    .forEach(file -> {
+                        String fileName = file.getName().replace(entry.getName(), "");
+                        File newFile = new File(targetFilePath + File.separator + fileName);
+                        try {
+                            FileUtils.copyInputStreamToFile(jar.getInputStream(file), newFile);
+                        } catch (IOException e) {
+                            log.error("Unable to get input stream for this file {}", targetFilePath + File.separator + fileName);
```
Comment (AmyKosinova) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: just log IO exception? Shouldn't we re-throw unchecked?

Comment (Hebcak) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Fixed it (I hope :-) )

Comment (AmyKosinova) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Well yea... It is always good to have it logged, exception may be not obvious enough to some users :smile: 
What about lines `172`, `184` and  `212`? Seems same to me.

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java (hunk 13 above)
Comment (Hebcak) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Line `172` is thrown inside the `log.error` on 173. We did it like that with @paveljandejsek. However I am not entirely sure, what's the difference between doing it this way, and the way I dealt with this one (`throw new UncheckedIOException(e);`)

Comment (paveljandejsek) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Not thrown, just logged with stacktrace.... So the flow will continue if exception is thrown, we catch it and dont rethrow it. I would probably add the rethrow to those lines as well, like Amy said.

Comment (Hebcak) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: So when we log it, it is just written out, but when we throw it, it stops the program?

Comment (AmyKosinova) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Maybe even better:
`copyResourceFromFile(Resource resource, File targetFile) throws IoException` as it is called from try block above. And same with `copyResourceFromJar(...)`
We can have single place handling same issue.

Comment (Hebcak) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Like this? Seems like an elegant solution.
```suggestion
                                      if (jarFile.isFile()) {
                copyResourceFromJar(resource, targetFilePath, targetFile, jarFile);
            } else {
                copyResourceFromFile(resource, targetFile);
            }
        } catch (IOException e) {
            log.error("Cannot copy resource file to target");
            throw new UncheckedIOException(e);
        }
    }

    private void copyResourceFromFile(Resource resource, File targetFile) throws IOException{
            if (resource.isFile() && resource.getFile().isDirectory()) {
                FileUtils.copyDirectory(resource.getFile(), targetFile);
            } else {
                FileUtils.copyInputStreamToFile(resource.getInputStream(), targetFile);
            }
    }

    private void copyResourceFromJar(Resource resource, String targetFilePath, File targetFile, File jarFile) throws IOException{
            try (JarFile jar = new JarFile(jarFile)) {
                JarEntry entry = jar.getJarEntry(resource.getFilename());
                if (entry.isDirectory()) {
                    jar.stream()
                        .filter(file -> file.getName().startsWith(entry.getName()) && file.getName().endsWith(".xslt"))
                        .forEach(file -> {
                            String fileName = file.getName().replace(entry.getName(), "");
                            File newFile = new File(targetFilePath + File.separator + fileName);
                            try {
                                FileUtils.copyInputStreamToFile(jar.getInputStream(file), newFile);
                            } catch (IOException e) {
                                log.error("Unable to get input stream for this file {}", targetFilePath + File.separator + fileName);
                                throw new UncheckedIOException(e);
                            }
                        });
                } else {
                    log.warn("File name: {}", entry.getName());

                    FileUtils.copyInputStreamToFile(jar.getInputStream(entry), targetFile);
                }
            }
    }
```

Comment (AmyKosinova) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: Like this. Maybe without fancy logging when throwing, but on the other hand little hipster parts makes code juicy :wink:

Comment (Hebcak) [reply jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: The log is just high-level overview of what's wrong. It is there to help with the first step of finding the root cause of the problem, but it is redundant.

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 14]
```diff
@@ -157,9 +160,47 @@ private void copyResourceToReportDirectory(Resource resource) {
...
+
+    private void copyResourceFromFile(Resource resource, File targetFile) throws IOException {
+        if (resource.isFile() && resource.getFile().isDirectory()) {
+            FileUtils.copyDirectory(resource.getFile(), targetFile);
+        } else {
+            FileUtils.copyInputStreamToFile(resource.getInputStream(), targetFile);
+        }
+    }
+
+    private void copyResourceFromJar(Resource resource, String targetFilePath, File targetFile, File jarFile) throws IOException {
```
Comment (AmyKosinova) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: `targetFilePath` and `targetFile` are effectively the same, can be reduced to one.

File: jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java [hunk 15]
```diff
@@ -157,9 +160,47 @@ private void copyResourceToReportDirectory(Resource resource) {
...
+        try (JarFile jar = new JarFile(jarFile)) {
+            JarEntry entry = jar.getJarEntry(resource.getFilename());
+            if (entry.isDirectory()) {
+                jar.stream()
+                    .filter(file -> file.getName().startsWith(entry.getName()) && file.getName().endsWith(".xslt"))
+                    .forEach(file -> {
+                        String fileName = file.getName().replace(entry.getName(), "");
+                        File newFile = new File(targetFilePath + File.separator + fileName);
+                        try {
+                            FileUtils.copyInputStreamToFile(jar.getInputStream(file), newFile);
```
Comment (AmyKosinova) [inline jbehave-support-core/src/main/java/org/jbehavesupport/core/report/XmlReporterFactory.java]: In this case stream brings more problems than improves readability. I suggest to refactor it to iterator approach.

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Comment by paveljandejsek** [comment]
Generated [reports](http://95.168.205.146:8080/reports/148/index.xml)

**Pull request merged**